cli/ss2.py

```diff
@@ -0,0 +1,317 @@
+
+    def __call__(self, nl_diag_sk):
+        raise RuntimeError('not implemented')
+
+
+class TCP(Protocol):
+
+    INET_DIAG_MEMINFO    = 1
+    INET_DIAG_INFO       = 2
+    INET_DIAG_VEGASINFO  = 3
+    INET_DIAG_CONG       = 4
+
+    def __init__(self, fmt='json'):
+        fmter = "_fmt_%s" % fmt
+        self._fmt = getattr(self, fmter, None)
+
+        if self._fmt == None:
+            raise RuntimeError("unknown formatting")
+
+        IDIAG_EXT_FLAGS = [ self.INET_DIAG_MEMINFO, 
+                            self.INET_DIAG_INFO, 
+                            self.INET_DIAG_VEGASINFO, 
+                            self.INET_DIAG_CONG ]
+
+        self.ext_f = 0
+        for f in IDIAG_EXT_FLAGS:
+            self.ext_f |= (1<<(f-1))
+
+    def __call__(self, nl_diag_sk):
+        sstats = nl_diag_sk.get_sock_stats(family=AF_INET, extensions=self.ext_f)
+        refined_stats = self._refine_diag_raw(sstats)
+        printable = self._fmt(refined_stats)
+
+        print printable
+    
+    def _refine_diag_raw(self, raw_stats):
+
+        refined = { 'TCP' : { 'flows' : [] } }
+
+        idiag_refine_map = { 
+                            'dst' :      'idiag_src',
+                            'src' :      'idiag_dst',
+                            'src_port' : 'idiag_sport',
+                            'dst_port' : 'idiag_dport',
+                            'inode' : 'idiag_inode',
+                            'iface_idx' :'idiag_if',
+                            'retrans' :  'idiag_retrans'
+                           }
+
+        for raw_flow in raw_stats:
+            vessel = {}
+            for k1,k2 in idiag_refine_map.items():
+                vessel[k1] = raw_flow[k2]
+
+            for ext_bundle in raw_flow['attrs']:
+                vessel = self._refine_extension(vessel, ext_bundle)
+            
+            refined['TCP']['flows'].append(vessel)
+
+        return refined
+
+    def _refine_extension(self, vessel, raw_ext):
+        k, content = raw_ext
+        ext_refine_map = {
+                          'meminfo' : {
+                                     'r' : 'idiag_rmem',
+                                     'w' : 'idiag_wmem',
+                                     'f' : 'idiag_fmem',
+                                     't' : 'idiag_tmem'
+                                    }
+                         }
+
+        if k == 'INET_DIAG_MEMINFO':
+            mem_k = 'meminfo'
+            vessel[mem_k] = {}
+            for k1, k2 in ext_refine_map[mem_k].items():
+                vessel[mem_k][k1] = content[k2]
+
+        elif k == 'INET_DIAG_CONG':
+            vessel['cong_algo'] = content
+
+        elif k == 'INET_DIAG_INFO':
+            vessel = self._refine_tcp_info(vessel, content)
+
+        elif k == 'INET_DIAG_SHUTDOWN':
+            pass
+
+        return vessel
+
+    def _refine_tcp_info(self, vessel, tcp_info_raw):
+        # normalizer
+        def rto_n_cb(key, value):
```

Comment:
I know this is still a draft, but if performance matters, I'd suggest moving nested `def`s elsewhere -- these will recreate a new function object each time the parent is called.

Reply:
It's a question of order of magnitude, still, this must performant and so you're right.